pyro/infer/mcmc/util.py

```diff
@@ -0,0 +1,137 @@
+            self._populate_cache(c, ordinal, enum_dims)
+
+    def _compute_log_prob_terms(self, model_trace):
+        """
+        Computes the conditional probabilities for each of the sites
+        in the model trace, and stores the result in `self._log_probs`.
+        """
+        model_trace.compute_log_prob()
+        ordering = {}
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample":
+                if len(site["cond_indep_stack"]) == 0:
+                    ordering[name] = self.root
+                else:
+                    ordering[name] = frozenset(tuple(self.root) + site["cond_indep_stack"])
+
+        # Collect log prob terms per independence context.
+        log_probs = defaultdict(list)
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample":
+                if is_validation_enabled():
+                    check_site_shape(site, self.max_iarange_nesting)
+                log_probs[ordering[name]].append(site["log_prob"])
+
+        for ordinal in log_probs:
+            self._log_probs[ordinal] = sum(log_probs[ordinal])
+
+    def _reduce(self, ordinal, agg_log_prob=torch.tensor(0.)):
+        """
+        Reduce the log prob terms for the given ordinal:
+          - taking log_sum_exp of factors in enum dims (i.e.
+            adding up the probability terms).
+          - summing up the dims within `max_iarange_nesting`.
+            (i.e. multiplying probs within independent batches).
+
+        :param ordinal: node (ordinal)
+        :param torch.Tensor agg_log_prob: aggregated `log_prob`
+            terms from the downstream nodes.
+        :return: `log_prob` with marginalized `iarange` and `enum`
+            dims.
+        """
+        log_prob = self._log_probs[ordinal] + agg_log_prob
+        for enum_dim in self._enum_dims[ordinal]:
+            log_prob = log_sum_exp(log_prob, dim=enum_dim, keepdim=True)
+        for marginal_dim in self._iarange_dims[ordinal]:
+            log_prob = log_prob.sum(dim=marginal_dim, keepdim=True)
+        return log_prob
+
+    def _aggregate_log_probs(self, ordinal):
+        """
+        Aggregate the `log_prob` terms using depth first search.
+        """
+        if self._children[ordinal] is None:
+            return self._reduce(ordinal)
+        agg_log_prob = torch.tensor(0.)
+        for c in self._children[ordinal]:
+            agg_log_prob = agg_log_prob + self._aggregate_log_probs(c)
+        return self._reduce(ordinal, agg_log_prob)
+
+    def log_prob(self, model_trace):
+        """
+        Returns the log pdf of `model_trace` by appropriately handling
+        enumerated log prob factors.
+
+        :return: log pdf of the trace.
+        """
+        if not self.has_enumerable_sites:
+            return model_trace.log_prob_sum()
+        self._compute_log_prob_terms(model_trace)
+        return self._aggregate_log_probs(self.root).sum()
```

Comment:
Do I understand correctly that:
1. all enumeration dims are to the left of `-max_iarange_nesting`, and they will have already been summed out by `self._aggregate_log_probs`
2. the result of `self._aggregate_log_probs(self.root)` will have non-reduced dims in exactly the global iaranges, i.e. the iaranges inside of which all sample statements are called, e.g. parallel particle iaranges?